nerd-dictation: init at unstable-2022-07-12

jtojnar opened the pull request:
###### Description of changes

Depends on https://github.com/NixOS/nixpkgs/pull/236313

Fixes: https://github.com/NixOS/nixpkgs/issues/176121

cc @guy-who-googles

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

jtojnar committed:
vosk-api: init at 0.3.45

Co-Authored-By: Reuben Steenekamp <reuben.steenekamp@gmail.com>

jtojnar committed:
vosk-python: init 0.3.45

Co-Authored-By: Reuben Steenekamp <reuben.steenekamp@gmail.com>

jtojnar committed:
nerd-dictation: init at unstable-2023-04-19

Co-Authored-By: Reuben Steenekamp <reuben.steenekamp@gmail.com>

jtojnar commented:
Currently broken at runtime:

> OSError: cannot load library '/nix/store/wc8vqabnp5rzdb93hqgjmcj057sbk7xf-vosk-api/lib/libvosk.so': /nix/store/wc8vqabnp5rzdb93hqgjmcj057sbk7xf-vosk-api/lib/libvosk.so: undefined symbol: _ZTVN3fst11SymbolTableE

jtojnar commented:
Appears to work now.

jtojnar commented:
I do not like how it is rewriting the existing text, trying https://github.com/NixOS/nixpkgs/pull/185255 as an alternative.

SuperSandro2000 commented on pkgs/applications/accessibility/nerd-dictation/default.nix:
```diff
@@ -0,0 +1,51 @@
+  '';
+
+  # No tests.
+  doCheck = false;
```
pythonImportsCheck missing

jtojnar replied:
This is an app not a library so not sure what would we even import.

SuperSandro2000 commented on pkgs/applications/accessibility/nerd-dictation/default.nix:
```diff
@@ -0,0 +1,51 @@
+  # No tests.
+  doCheck = false;
+
+  # Broken for some reason.
```
Can you paste the log?

jtojnar replied:
```
pythonCatchConflictsPhase
Found duplicated packages in closure for dependency 'nerd-dictation': 
  nerd-dictation 0.0.0 (/nix/store/3dc8q0l76p27nnyv8b8aijfhs6ssifk7-nerd-dictation-unstable-2022-07-12/lib/python3.10/site-packages)
  nerd-dictation 0.0.0 (/nix/store/3dc8q0l76p27nnyv8b8aijfhs6ssifk7-nerd-dictation-unstable-2022-07-12/lib/python3.10/site-packages)

Package duplicates found in closure, see above. Usually this happens if two packages depend on different version of the same dependency.
```

SuperSandro2000 replied:
> nerd_dictation-0.0.0.dist-info

The version number is not set correct.

SuperSandro2000 replied:
https://github.com/ideasman42/nerd-dictation/blob/master/package/python/setup.py#L43

The version number is hardcoded to ``0.0.0`` which is the default value and should be treated as an error.

SuperSandro2000 replied:
The layout in site-packages is really strange and very custom.
Anything installed under site-packages should be importable and it is frowned upon to files directly under site-packages.

I expect this to cause more trouble in the future.

Compared to a standard python package:

```
 ➜ tree result
 result ⇒ /nix/store/3qfd3gwpar83k89wdnw8pab7xr3bfv52-nerd-dictation-unstable-2022-07-12
├──  bin
│   └──  nerd-dictation
├──  lib
│   └──  python3.10
│       └──  site-packages
│           ├──  __pycache__
│           │   └──  setup.cpython-310.pyc
│           ├──  doc
│           │   └──  readme.rst
│           ├──  nerd-dictation
│           ├──  nerd_dictation-0.0.0.dist-info
│           │   ├──  direct_url.json
│           │   ├──  INSTALLER
│           │   ├──  METADATA
│           │   ├──  RECORD
│           │   ├──  REQUESTED
│           │   ├──  top_level.txt
│           │   └──  WHEEL
│           ├──  nerd_dictation.egg-info
│           │   ├──  dependency_links.txt
│           │   ├──  PKG-INFO
│           │   ├──  requires.txt
│           │   ├──  SOURCES.txt
│           │   └──  top_level.txt
│           ├──  nix_run_setup
│           └──  setup.py
├──  nerd-dictation
│   ├──  nerd-dictation
│   └──  readme.rst
└──  nix-support
    └──  propagated-build-inputs
```

```
 ➜ tree result
 result ⇒ /nix/store/k38smakhffi54jgpj21d4l51w5va85ly-python3.10-jsonschema-4.7.2
├──  bin
│   └──  jsonschema
├──  lib
│   └──  python3.10
│       └──  site-packages
│           ├──  jsonschema
│           │   ├──  __init__.py
│           │   ├──  __main__.py
│           │   ├──  __pycache__
│           │   │   ├──  __init__.cpython-310.pyc
│           │   │   ├──  __main__.cpython-310.pyc
│           │   │   ├──  _format.cpython-310.pyc
│           │   │   ├──  _legacy_validators.cpython-310.pyc
│           │   │   ├──  _reflect.cpython-310.pyc
│           │   │   ├──  _types.cpython-310.pyc
│           │   │   ├──  _utils.cpython-310.pyc
│           │   │   ├──  _validators.cpython-310.pyc
│           │   │   ├──  cli.cpython-310.pyc
│           │   │   ├──  exceptions.cpython-310.pyc
│           │   │   ├──  protocols.cpython-310.pyc
│           │   │   └──  validators.cpython-310.pyc
│           │   ├──  _format.py
│           │   ├──  _legacy_validators.py
│           │   ├──  _reflect.py
│           │   ├──  _types.py
│           │   ├──  _utils.py
│           │   ├──  _validators.py
│           │   ├──  benchmarks
│           │   │   ├──  __init__.py
│           │   │   ├──  __pycache__
│           │   │   │   ├──  __init__.cpython-310.pyc
│           │   │   │   ├──  issue232.cpython-310.pyc
│           │   │   │   └──  json_schema_test_suite.cpython-310.pyc
│           │   │   ├──  issue232
│           │   │   │   └──  issue.json
│           │   │   ├──  issue232.py
│           │   │   └──  json_schema_test_suite.py
│           │   ├──  cli.py
│           │   ├──  exceptions.py
│           │   ├──  protocols.py
│           │   ├──  schemas
│           │   │   ├──  draft2019-09.json
│           │   │   ├──  draft2020-12.json
│           │   │   ├──  draft3.json
│           │   │   ├──  draft4.json
│           │   │   ├──  draft6.json
│           │   │   ├──  draft7.json
│           │   │   └──  vocabularies.json
│           │   ├──  tests
│           │   │   ├──  __init__.py
│           │   │   ├──  __pycache__
│           │   │   │   ├──  __init__.cpython-310.pyc
│           │   │   │   ├──  _helpers.cpython-310.pyc
│           │   │   │   ├──  _suite.cpython-310.pyc
│           │   │   │   ├──  fuzz_validate.cpython-310.pyc
│           │   │   │   ├──  test_cli.cpython-310.pyc
│           │   │   │   ├──  test_deprecations.cpython-310.pyc
│           │   │   │   ├──  test_exceptions.cpython-310.pyc
│           │   │   │   ├──  test_format.cpython-310.pyc
│           │   │   │   ├──  test_jsonschema_test_suite.cpython-310.pyc
│           │   │   │   ├──  test_types.cpython-310.pyc
│           │   │   │   ├──  test_utils.cpython-310.pyc
│           │   │   │   └──  test_validators.cpython-310.pyc
│           │   │   ├──  _helpers.py
│           │   │   ├──  _suite.py
│           │   │   ├──  fuzz_validate.py
│           │   │   ├──  test_cli.py
│           │   │   ├──  test_deprecations.py
│           │   │   ├──  test_exceptions.py
│           │   │   ├──  test_format.py
│           │   │   ├──  test_jsonschema_test_suite.py
│           │   │   ├──  test_types.py
│           │   │   ├──  test_utils.py
│           │   │   └──  test_validators.py
│           │   └──  validators.py
│           └──  jsonschema-4.7.2.dist-info
│               ├──  direct_url.json
│               ├──  entry_points.txt
│               ├──  INSTALLER
│               ├──  licenses
│               │   └──  COPYING
│               ├──  METADATA
│               ├──  RECORD
│               ├──  REQUESTED
│               └──  WHEEL
└──  nix-support
    └──  propagated-build-inputs
```

jtojnar replied:
Looks like there is something here: https://github.com/ideasman42/nerd-dictation/pull/79

SuperSandro2000 commented on pkgs/development/python-modules/vosk-python/default.nix:
```diff
@@ -0,0 +1,48 @@
+      src = ./fix-paths.patch;
+      vosk_api = vosk-api;
+    })
+  ];
```
```suggestion
  ];

  postPatch = ''
    cd python
  '';
```

SuperSandro2000 commented on pkgs/development/python-modules/vosk-python/fix-paths.patch:
```diff
@@ -0,0 +1,64 @@
+-    elif sys.platform == 'darwin':
+-        return _ffi.dlopen(os.path.join(dlldir, "libvosk.dyld"))
++    if sys.platform == 'linux':
++        return _ffi.dlopen("@vosk_api@/lib/libvosk.so")
```
Then we should set meta.platforms to linux.

SuperSandro2000 commented on pkgs/tools/audio/kaldi/default.nix:
```diff
@@ -23,22 +23,22 @@ let
 stdenv.mkDerivation {
   pname = "kaldi";
-  version = "2021-12-03";
+  version = "2022-07-20";
```
```suggestion
  version = "unstable-2022-07-20";
```

SuperSandro2000 commented on pkgs/tools/audio/vosk-api/default.nix:
```diff
@@ -0,0 +1,60 @@
+  ];
+
+  buildInputs = [
+    (kaldi.overrideAttrs (attrs: {
```
Please move this to top-level (preferred) or in a let in

jtojnar replied:
I have yet to see applying patches in top-level. And do not see any benefit to extending the scope either.

SuperSandro2000 replied:
We also almost never apply patches in the inputs itself. Also the unpatched kaldi version could easily be used somewhere else in the file by accident.

jtojnar replied:
The only way to prevent that is to use shadowing. So I went there.

jtojnar replied:
Top-level overrides are now discouraged: https://github.com/NixOS/nixpkgs/issues/204303

reivilibre commented:
@jtojnar I'd be interested in having `vosk-api` packaged in NixOS if nothing else; assuming you're too busy to continue with this PR at the moment(?), would it be acceptable to you if I took your work packaging `vosk-api` and opened a PR with just that?

jtojnar commented:
On Wed, Sep 28 2022 at 15:31:41 -07:00:00, reivilibre 
***@***.***> wrote:
> @jtojnar I'd be interested in having vosk-api packaged in NixOS if 
> nothing else; assuming you're too busy to continue with this PR at 
> the moment(?), would it be acceptable to you if I took your work 
> packaging vosk-api and opened a PR with just that?
> 

Feel free, but IIRC there is already another PR for that.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/speech-to-text-virtual-keyboard-in-nixos/22684/2

jtojnar committed:
kaldi: Fix build

It looks like after a bump, CMake started passing `--git-dir=.git` as the first argument of `git rev-parse`.
But we do not really need to spoof `git` program now that Kaldi uses CMake’s `FetchContent` module.
We can just pass the path using configure flag directly:

https://cmake.org/cmake/help/latest/module/FetchContent.html#variable:FETCHCONTENT_SOURCE_DIR_%3CuppercaseName%3E

Git was also inadvertently used for generating `KALDI_PATCH_NUMBER`
in `VersionHelper`, to be potentially appended to `KALDI_VERSION`
but it was not actually being done by default.
We can set the variable directly to skip `VersionHelper`.

Also remove redundant `enableParallelBuild` since CMake enables it.

And drop separate `dev` output since Kaldi currently hardcodes
the `include/` directory so consumers would not be able to find them.
Splitting it out only removes 7 out of 303 MB.

jtojnar committed:
kaldi: Remove openfst from inputs

It is shadowed by fetchgit source and can only be used vendored.

jtojnar committed:
kaldi: Add update script

Also switch to finalAttrs so that the values can be easily overridden.

jtojnar committed:
kaldi: unstable-2022-09-26 → unstable-2023-05-02

https://github.com/kaldi-asr/kaldi/compare/f6f4ccaf213f0fe8b26e633a7dc0c802150626a0..71f38e62cad01c3078555bfe78d0f3a527422d75

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/assistance-needed-in-packaging-python-program/33420/3
